### PR TITLE
App Services Base + Media + Weather + NAV

### LINKED ISSUE #145: SDL 0167 - App Services - Weather Service
Proposal: [SDL 0167 - App Services - Weather Service](https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0167-app-services.md#weather)

> SmartDeviceLink has always been limited to a 1 - 1 relationship between module and smartphone apps. However, to create truly compelling experiences this should be expanded. This proposal will aim to introduce a paradigm that allows apps to advertise a "service" they offer that the module and other apps can leverage. This allows users to augment their vehicle's IVI system with off-board solutions. This proposal also details how SDL apps can communicate with each other through the module.

https://github.com/smartdevicelink/sdl_evolution/issues/581

Steering Committee Decision:

> The Steering Committee voted to accept the Weather Service type section of the larger App Services proposal, with the revisions included in [this comment](https://github.com/smartdevicelink/sdl_evolution/issues/581#issuecomment-412519648).

The accepted revisions were made on 2019-01-24.

### PR DESCRIPTION
**Pull request by Jack-Byrne**
Fixes #145 #146

### CODE REVIEW

File: MOBILE_API.xml [hunk 1]
```diff
@@ -2608,6 +2608,10 @@
         <element name="UnsubscribeWayPointsID" value="47" hexvalue="2F" since="4.1" />
         <element name="GetSystemCapabilityID" value="48" hexvalue="30" since="4.5" />
         <element name="SendHapticDataID" value="49" hexvalue="31" since="4.5" />
+        <element name="PublishAppServiceID" value="52" hexvalue="34" since="5.1" />
+        <element name="GetAppServiceDataID" value="53" hexvalue="35" since="5.1" />
+        <element name="GetFileID" value="54" hexvalue="36" since="5.1" />
+        <element name="PerformAppServiceInteractiontFileID" value="55" hexvalue="37" since="5.1" />
```
Comment (joeygrover) [inline MOBILE_API.xml]: Typo, shouldn't have `tFile` attached to name should be `PerformAppServiceInteractionID`

File: MOBILE_API.xml [hunk 2]
```diff
@@ -3699,6 +3709,417 @@
...
+    <struct name="NavigationServiceManifest">
+		<param name="acceptsWayPoints" type="Boolean" mandatory="false">
+		<description>  Informs the subscriber if this service can actually accept way points.</description>
+		</param>
+	</struct>
+
+    <struct name="NavigationServiceData">
+	    <description> This data is related to what a navigation service would provide.</description>
+	
+	    <param name="timestamp" type="" mandatory="true">
```
Comment (jordynmackool) [inline MOBILE_API.xml]: Type should be `DateTime`

**Review by jordynmackool — CHANGES_REQUESTED** [review]

File: MOBILE_API.xml [hunk 3]
```diff
@@ -3699,6 +3709,417 @@
...
+            </description>
+        </param>
+        <param name="trackPlaybackDuration" type="Int" mandatory="false">
+            <description>
+            Music: The total duration of the track in seconds
+            Podcast: The total duration of the episode in seconds
+            Audiobook: The total duration of the current segment (e.g. the chapter) in seconds
+            </description>
+        </param>
+        <param name="queuePlaybackProgess" type="Int" mandatory="false">
```
Comment (jordynmackool) [inline MOBILE_API.xml]: Update `queuePlaybackProgess` to `queuePlaybackProgress`

**Review by jordynmackool — CHANGES_REQUESTED** [review]

File: MOBILE_API.xml [hunk 4]
```diff
@@ -3699,6 +3709,417 @@
...
+        <element name="BIFURCATION">
+            <description> A junction where the road splits off into two paths; a fork in the road.</description>
+        </element>
+        <element name="MULTI_CARRIAGEWAY">
+            <description> A junction that has multiple intersections and paths.</description>
+        </element>              
+        <element name="ROUNDABOUT">
+            <description>A junction where traffic moves in a single direction around a central, non-traversable point to reach one of the connecting roads.</description>
+        </element>
+        <element name="TRAVESABLE_ROUNDABOUT">
```
Comment (jordynmackool) [inline MOBILE_API.xml]: Update to "TRAVERSABLE_ROUNDABOUT"

**Review by jordynmackool — CHANGES_REQUESTED** [review]

File: MOBILE_API.xml (hunk 2 above)
Comment (jordynmackool) [reply MOBILE_API.xml]: Param name should be `timeStamp`

Updates should look like: 
`<param name="timeStamp" type="DateTime" mandatory="true">
                <description> This is the timestamp of when the data was generated. This is to ensure any time or distance given in the data can accurately be adjusted if necessary. </description>`
            </param>

**Review by jordynmackool — CHANGES_REQUESTED** [review]
@JackLivio Please see noted comment

File: MOBILE_API.xml [hunk 5]
```diff
@@ -6606,6 +7022,152 @@
...
+            
+    </function>
+
+    <function name="PerformAppServiceInteraction" functionID="PerformAppServiceInteractionID" messagetype="request" since="5.1">
+
+        <param name="serviceUri" type="String"  mandatory="true">
+            <description>Fully qualified URI based on the URI prefix and URI scheme the app service provided. SDL makes no gurantee that this URI is correct.</description>
+        </param>
+        
+        <param name="appServiceID" type="String" mandatory="true">
```
Comment (jordynmackool) [inline MOBILE_API.xml]: The proposal shows `param name="serviceID"` not `appServiceID`

param name="serviceID" type="String" mandatory="true">
		<description> The service ID that the app consumer wishes to send this URI.</description>

**Review by jordynmackool — CHANGES_REQUESTED** [review]

File: MOBILE_API.xml [hunk 6]
```diff
@@ -3699,7 +3688,428 @@
...
+        <param name="eta" type="DateTime" mandatory="false"/>
+
+        <param name="bearing" type="Integer" minvalue="0" maxvalue="359" mandatory="false">
+            <description> The angle at which this instruction takes place. For example, 0 would mean straight, less than 45 is bearing right, greater than 135 is sharp right, between 45 and 135 is a regular right, and 180 is a U-Turn, etc. </description>
+        </param>
+
+        <param name="junctionType" type="NavigationJunction" mandatory="false"/>
+
+        <param name="drivingSide" type="Direction" mandatory="false">
+            <description> Used to infer which side of the road this instruction takes place. For a U-Turn (Action=Turn, direction=180) this will determine which direction the turn should take place. </description>
```
Comment (NicoleYarroch) [inline MOBILE_API.xml]: * _Action=Turn_ should be formatted like so: _action=TURN_
* _direction=180_ should be _bearing=180_

File: MOBILE_API.xml [hunk 7]
```diff
@@ -3699,7 +3688,428 @@
...
+            <description> If true, the service is published and available. If false, the service has likely just been unpublished, and should be considered unavailable.</description>
+        </param>
+
+        <param name="serviceActive" type="Boolean" mandatory="true">
+            <description> If true, the service is the active primary service of the supplied service type. It will receive all potential RPCs that are passed through to that service type. If false, it is not the primary service of the supplied type. See servicePublished for its availability. </description>
+        </param>
+    </struct>
+
+    <struct name="AppServiceData" since="5.1">
+        <description> Contains all the current data of the app service. The serviceType will link to which of the service data objects are included in this object. (eg if service type equals MEDIA, the mediaServiceData param should be included.</description>
```
Comment (NicoleYarroch) [inline MOBILE_API.xml]: _object. (eg if service type equals MEDIA, the mediaServiceData param should be included._
The text above has some formatting issues. It should be:
_object (e.g. if the service type is MEDIA, the mediaServiceData param should be included)._

File: MOBILE_API.xml [hunk 8]
```diff
@@ -3699,7 +3688,428 @@
...
+            <description>An array of currently available services. If this is an update to the capability the affected services will include an update reason in that item</description>
+        </param>
+    </struct>
+
+    <!-- End App Services -->
     
+    <struct name="SystemCapability" since="4.5">
+        <description>The systemCapabilityType identifies which data object exists in this struct. For example, if the SystemCapability Type is NAVIGATION then a "navigationCapability" should exist</description>
+        <param name="systemCapabilityType" type="SystemCapabilityType" mandatory="true">
+            <description>Used as a descriptor of what data to expect in this struct. The corresponding param to this enum should be included and the only other para included.</description>
```
Comment (NicoleYarroch) [inline MOBILE_API.xml]: _para_ should be _param_

File: MOBILE_API.xml [hunk 9]
```diff
@@ -6013,7 +6423,65 @@
...
+            <description>Provides additional human readable info regarding the result.</description>
+        </param>
+
+        <!-- Specific response data -->
+        <param name="offset" type="Integer" minvalue="0" maxvalue="2000000000" mandatory="false">
+            <description>Optional offset in bytes for resuming partial data chunks</description>
+        </param>
+
+        <param name="length" type="Integer" minvalue="0" maxvalue="2000000000" mandatory="false">
+            <description>Optional length in bytes for resuming partial data chunks if offset is set to 0, then length is the total length of the file to be downloaded</description>
```
Comment (NicoleYarroch) [inline MOBILE_API.xml:6527]: The description is a run-on sentence. 
It should be: 
_Optional length in bytes for resuming partial data chunks. If offset is set to 0, then length is the total length of the file to be downloaded_

File: MOBILE_API.xml [hunk 10]
```diff
@@ -6606,7 +7077,118 @@
...
+		    <description>The type of service that is to be offered by this app. See AppServiceType for known enum equivalent types. Parameter is a string to allow for new service types to be used by apps on older versions of SDL Core.</description>
+	    </param>
+        
+        <param name="subscribe" type="Boolean" mandatory="false">
+            <description> If true, the consumer is requesting to subscribe to all future updates from the service publisher. If false, the consumer doesn't wish to subscribe and should be unsubscribed if it was previously subscribed.</description>
+        </param>
+    </function>
+
+    <function name="GetAppServiceData" functionID="GetAppServiceDataID" messagetype="response" since="5.1">
+        <description> This response includes the data that is requested from the specific service</description>
```
Comment (NicoleYarroch) [inline MOBILE_API.xml]: _This response includes the data that is requested from the specific service_
should be 
_This response includes the data that was requested from the specific service_

File: MOBILE_API.xml [hunk 11]
```diff
@@ -6606,7 +7077,118 @@
...
+            <description>Provides additional human readable info regarding the result.</description>
+        </param>
+        
+        <!-- Specific response data -->        
+        <param name="serviceData" type="AppServiceData" mandatory="false"/>            
+    </function>
+
+    <function name="PerformAppServiceInteraction" functionID="PerformAppServiceInteractionID" messagetype="request" since="5.1">
+        <param name="serviceUri" type="String"  mandatory="true">
+            <description>Fully qualified URI based on the URI prefix and URI scheme the app service provided. SDL makes no gurantee that this URI is correct.</description>
```
Comment (NicoleYarroch) [inline MOBILE_API.xml]: _gurantee_ is misspelled. Should be _guarantee_

Comment (NicoleYarroch) [reply MOBILE_API.xml]: Weren't the URI prefix and URI scheme parameters removed from the app service manifest? 
> Fully qualified URI based on the URI prefix and URI scheme the app service provided

So this documentation would be incorrect.

File: MOBILE_API.xml [hunk 12]
```diff
@@ -6606,7 +7077,118 @@
...
+        <param name="resultCode" type="Result" platform="documentation" mandatory="true">
+            <description>See Result. All results will be available for this response.</description>
+        </param>
+        
+        <param name="info" type="String" maxlength="1000" mandatory="false" platform="documentation">
+            <description>Provides additional human readable info regarding the result.</description>
+        </param>
+
+        <param name="serviceSpecificResult" type="String" mandatory="false">
+            <description>The service can provide specific result strings to the consumer through this param. These results should be described in the URI schema set in the Service Manifest</description>
```
Comment (NicoleYarroch) [inline MOBILE_API.xml]: > These results should be described in the URI schema set in the Service Manifest

Is this correct now that the URI prefix and URI scheme parameters were removed from the app service manifest?

**Review by jordynmackool — APPROVED** [review]

File: MOBILE_API.xml (hunk 6 above)
Comment (jacobkeeler) [reply MOBILE_API.xml]: Fixed in https://github.com/smartdevicelink/rpc_spec/pull/149/commits/06a3394c636aa24d9b880a352c67c94700476e0e

File: MOBILE_API.xml (hunk 7 above)
Comment (jacobkeeler) [reply MOBILE_API.xml]: Fixed in https://github.com/smartdevicelink/rpc_spec/pull/149/commits/06a3394c636aa24d9b880a352c67c94700476e0e

File: MOBILE_API.xml (hunk 8 above)
Comment (jacobkeeler) [reply MOBILE_API.xml]: Fixed in https://github.com/smartdevicelink/rpc_spec/pull/149/commits/06a3394c636aa24d9b880a352c67c94700476e0e

File: MOBILE_API.xml (hunk 9 above)
Comment (jacobkeeler) [reply MOBILE_API.xml:6527]: Fixed in https://github.com/smartdevicelink/rpc_spec/pull/149/commits/06a3394c636aa24d9b880a352c67c94700476e0e

File: MOBILE_API.xml (hunk 10 above)
Comment (jacobkeeler) [reply MOBILE_API.xml]: Fixed in https://github.com/smartdevicelink/rpc_spec/pull/149/commits/06a3394c636aa24d9b880a352c67c94700476e0e

File: MOBILE_API.xml (hunk 11 above)
Comment (jacobkeeler) [reply MOBILE_API.xml]: Fixed in https://github.com/smartdevicelink/rpc_spec/pull/149/commits/06a3394c636aa24d9b880a352c67c94700476e0e

File: MOBILE_API.xml (hunk 12 above)
Comment (jacobkeeler) [reply MOBILE_API.xml]: Fixed in https://github.com/smartdevicelink/rpc_spec/pull/149/commits/06a3394c636aa24d9b880a352c67c94700476e0e

File: MOBILE_API.xml [hunk 13]
```diff
@@ -3699,7 +3688,428 @@
...
+            <description> 0 to 1, percentage chance </description>
+        </param>
+        <param name="precipType" type="String" mandatory="false">
+            <description> e.g. "rain", "snow", "sleet", "hail" </description>
+        </param>
+        <param name="visibility" type="Float" mandatory="false">
+            <description> In km </description>
+        </param>
+
+        <param name="weatherIconImageName" type="String" mandatory="false"/>
```
Comment (jordynmackool) [inline MOBILE_API.xml]: Update so the name is `weatherIcon` and the data type is `Image`

File: README.md [hunk 14]
```diff
@@ -2629,6 +2693,218 @@ Defines haptic data for each user control object for video streaming application
...
+|`windGust`|Float|False|km/hr |
+|`windSpeed`|Float|False|km/hr |
+|`nearestStormBearing`|Integer|False|In degrees, true north at 0 degrees |
+|`nearestStormDistance`|Integer|False|In km |
+|`precipAccumulation`|Float|False|cm |
+|`precipIntensity`|Float|False|cm of water per hour |
+|`precipProbability`|Float|False|0 to 1, percentage chance |
+|`precipType`|String|False|e.g. "rain", "snow", "sleet", "hail" |
+|`visibility`|Float|False|In km |
+|`weatherIconImageName`|String|False||
```
Comment (jordynmackool) [inline README.md]: Update so the name is `weatherIcon` and the data type is `Image`

**Review by jordynmackool — CHANGES_REQUESTED** [review]

File: README.md (hunk 14 above)
Comment (jacobkeeler) [reply README.md]: Fixed 👍

File: MOBILE_API.xml (hunk 13 above)
Comment (jacobkeeler) [reply MOBILE_API.xml]: Fixed 👍

**Review by jordynmackool — APPROVED** [review]

File: MOBILE_API.xml [hunk 15]
```diff
@@ -3732,7 +3721,428 @@
...
+
+    <struct name="AppServiceManifest" since="5.1">
+        <description> This manifest contains all the information necessary for the service to be published, activated, and consumers able to interact with it </description>
+
+        <param name="serviceName" type="String" mandatory="false">
+            <description> Unique name of this service </description>
+        </param>
+
+        <param name="serviceType" type="String" mandatory="true">
+            <description> The type of service that is to be offered by this app </description>
```
Comment (theresalech) [inline MOBILE_API.xml]: The proposal also includes this in the description:
>See AppServiceType for known enum equivalent types. Parameter is a string to allow for new service types to be used by apps on older versions of SDL Core.</description>
		</param>

**Review by theresalech — APPROVED** [review]

**Pull request merged**